hail/python/hail/methods/pca.py

```diff
@@ -207,13 +207,261 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
     return hl.eval(g.eigenvalues), scores, None if t is None else t.drop('eigenvalues', 'scores')
 
 
-@typecheck(entry_expr=expr_float64,
+class TallSkinnyMatrix:
+    def __init__(self, block_table, block_expr, source_table, col_key):
+        self.col_key = col_key
+        first_block = block_expr.take(1)[0]
+        self.ncols = first_block.shape[1]
+        self.block_table = block_table
+        self.block_expr = block_expr
+        self.source_table = source_table
+
+
+def _make_tsm(entry_expr, block_size):
+    mt = matrix_table_source('_make_tsm/entry_expr', entry_expr)
+    A, ht = mt_to_table_of_ndarray(entry_expr, block_size, return_checkpointed_table_also=True)
+    A = A.persist()
+    return TallSkinnyMatrix(A, A.ndarray, ht, list(mt.col_key))
+
+
+def _make_tsm_from_call(call_expr, block_size, mean_center=False, hwe_normalize=False):
+    mt = matrix_table_source('_make_tsm/entry_expr', call_expr)
+    mt = mt.select_entries(__gt=call_expr.n_alt_alleles())
+    if mean_center or hwe_normalize:
+        mt = mt.annotate_rows(__AC=agg.sum(mt.__gt),
+                              __n_called=agg.count_where(hl.is_defined(mt.__gt)))
+        mt = mt.filter_rows((mt.__AC > 0) & (mt.__AC < 2 * mt.__n_called))
+
+        n_variants = mt.count_rows()
+        if n_variants == 0:
+            raise FatalError("_make_tsm: found 0 variants after filtering out monomorphic sites.")
+        info(f"_make_tsm: found {n_variants} variants after filtering out monomorphic sites.")
+
+        mt = mt.annotate_rows(__mean_gt=mt.__AC / mt.__n_called)
+        mt = mt.unfilter_entries()
+
+        mt = mt.select_entries(__x=hl.or_else(mt.__gt - mt.__mean_gt, 0.0))
+
+        if hwe_normalize:
+            mt = mt.annotate_rows(
+                __hwe_scaled_std_dev=hl.sqrt(mt.__mean_gt * (2 - mt.__mean_gt) * n_variants / 2))
+            mt = mt.select_entries(__x=mt.__x / mt.__hwe_scaled_std_dev)
+    else:
+        mt = mt.select_entries(__x=mt.__gt)
+
+    A, ht = mt_to_table_of_ndarray(mt.__x, block_size, return_checkpointed_table_also=True)
+    A = A.persist()
+    return TallSkinnyMatrix(A, A.ndarray, ht, list(mt.col_key))
+
+
+class KrylovFactorization:
+    # For now, all three factors are `NDArrayExpression`s, but eventually U and/or V should be
+    # allowed to be distributed. All properties must be persisted.
+    def __init__(self, U, R, V, k):
+        self.U = U
+        self.V = V
+        self.k = k
+        (self.U1, self.S, self.V1t) = hl.nd.svd(R, full_matrices=False)._persist()
+
+    def reduced_svd(self, k):
+        S = self.S[:k]._persist()
+
+        if self.U is None:
+            U = None
+        else:
+            U = (self.U @ self.U1[:, :k])._persist()
+
+        if self.V is None:
+            V = None
+        else:
+            V = (self.V @ self.V1t.T[:, :k])._persist()
+
+        return U, S, V
+
+    def spectral_moments(self, num_moments, R):
```

Comment:
Actually, one note. I think maybe you ought to return a single struct from this method instead of 2 separate expressions. You're making it very easy to accidentally do double work by forking the IR.

Reply:
In the place you use the method, you actually recombine the results into a struct, presumably for this reason, so why not just return a struct here.